stan/math/opencl/matrix_cl.hpp

```diff
@@ -213,14 +211,14 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
       check_opencl_error("copy (OpenCL)->(OpenCL)", e);
     }
   }
-
-  matrix_cl(matrix_cl<T>&& A)
-      : buffer_cl_(A.buffer_cl_),
+  template <typename Mat, require_matrix_cl_t<std::is_arithmetic, Mat>...>
+  explicit matrix_cl(Mat&& A)
+      : buffer_cl_(std::forward<decltype(A.buffer_cl_)>(A.buffer_cl_)),
```

Comment:
Dumb Q, if A is an rvalue are it's members still lvalues because they are named?

Reply:
Not sure. My gut feeling says they should be rvalues too.

Reply:
> These std::forward calls do the same as std::move would, since buffer_cl_, write_events_ and read_events_ are not templated.

They are the same as std::move if A is an rvalue, if it's an lvalue it won't move (reference collapsing)

Reply:
I think we are agreeing with what the code is doing but not the implications

I changed this here is to it handle both const and non-const lvalues and r values, is there a reason we want a separate constructor for all of them?

Reply:
Gaaahhh nah nvm I'm dumb and your right we need to explicitly run that `clCopyBuffer`